<!DOCTYPE html>
<html lang="en-us">
<head>
    <meta charset="utf-8">
    <!--<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">-->

    <title> About - EpiSurf</title>
    <meta name="description" content="">
    <meta name="author" content="">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <!-- Basic Styles -->
    <link rel="stylesheet" type="text/css" media="screen"
          href="./bootstrap.min.css">
    <link rel="stylesheet" href="inc/font-awesome.min.4.4.0.css">

    <!-- SmartAdmin Styles : Please note (smartadmin-production.css) was created using LESS variables -->
    <link rel="stylesheet" type="text/css" media="screen"
          href="smartadmin-production.min.css">
    <link rel="stylesheet" type="text/css" media="screen" href="contact.css">
    <link rel="icon" href="favicon.ico">

</head>
<body id="extr-page">
<!-- POSSIBLE CLASSES: minified, fixed-ribbon, fixed-header, fixed-width
     You can also add different skin classes such as "smart-skin-1", "smart-skin-2" etc...-->


<div id="main" role="main">
    <!-- MAIN CONTENT -->
    <div id="content0" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">How to cite</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                            A paper that describes EpiSurf is currently under preparation.
                            If you use EpiSurf or use screenshots generated with it
                            please do ensure you credit our website (providing a link back to  <a href="http://gmql.eu/episurf/" target="_blank">http://gmql.eu/episurf/</a>)
                            <br><br>
                            If you use VirusViz, please cite our (recently accepted) paper:<br>
                            <b>A. Bernasconi</b>, <b>A. Gulino</b>, T. Alfonsi, A. Canakoglu, P. Pinoli, A. Sandionigi, S. Ceri.
                            VirusViz: Comparative analysis and effective visualization of viral nucleotide and amino acid variants. <i>Accepted to Nucleic acids research - Web Server Issue 2021</i>.
                            <br><br>
                            If you use data extracted from ViruSurf application, you may cite:<br>
                            <b>A. Canakoglu</b>, <b>P. Pinoli</b>, A. Bernasconi, T. Alfonsi, D.P. Melidis, S. Ceri.
                            2021. ViruSurf: an integrated database to investigate viral sequences. <i>Nucleic acids research</i>, 49(D1), pp.D817-D824.
                            <a href="https://doi.org/10.1093/nar/gkaa846" target="_blank">https://doi.org/10.1093/nar/gkaa846</a>
                            <br><br>
                            If you used the data model, you may cite:<br>
                            A. Bernasconi, A. Canakoglu, P. Pinoli, S. Ceri. 2020, November.
                            Empowering Virus Sequence Research Through Conceptual Modeling.
                            In <i>International Conference on Conceptual Modeling</i> (pp. 388-402). Springer, Cham. <a href="https://doi.org/10.1007/978-3-030-62522-1_29" target="_blank">https://doi.org/10.1007/978-3-030-62522-1_29</a>
                            </td>
                        </tr>
                        </tbody>
                    </table>
                </div>
            </div>
        </div>
    </div>

    <div id="content1" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">Contacts</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Dr. Arif Canakoglu</b>
                                <p>Project Leadership, Concept Design</p>
                                <p>  <a href="mailto:arif.canakoglu@polimi.it">arif.canakoglu@polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Dr. Anna Bernasconi</b>
                                <p>Examples, Storytelling, Testing, Concept Design</p>
                                <p>  <a href="mailto:anna.bernasconi@polimi.it">anna.bernasconi@polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Luca Cilibrasi</b>
                                <p>Main Development</p>
                                <p>  <a href="mailto:luca.cilibrasi@mail.polimi.it">luca.cilibrasi@mail.polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Ruba Al Khalaf</b>
                                <p>Domain Knowledge Research, Examples</p>
                                <p>  <a href="mailto:arif.canakoglu@polimi.it">arif.canakoglu@polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Tommaso Alfonsi</b>
                                <p>Data Architect</p>
                                <p>  <a href="mailto:tommaso.alfonsi@polimi.it">tommaso.alfonsi@polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Prof. Stefano Ceri</b>
                                <p>Principal Investigator, Concept Design</p>
                                <p>  <a href="mailto:stefano.ceri@polimi.it">stefano.ceri@polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b style="font-size:1.3em">Dr. Pietro Pinoli</b>
                                <p>Concept Design</p>
                                <p>  <a href="mailto:pietro.pinoli@polimi.it">pietro.pinoli@polimi.it</a></p>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <b>Dipartimento di Elettronica, Informazione e Bioingegneria</b><br>
                                Politecnico di Milano<br>
                                Via Ponzio 34/5 Milano<br>
                                20133 Milano<br>
                                Italy
                            </td>
                        </tr>

                        </tbody>
                    </table>
                </div>
            </div>
        </div>
    </div>


<div id="content1" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">Third-party software</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                                <p><a href="https://www.postgresql.org/">PostgreSQL</a></p>
                                <p><a href="https://vuejs.org/">Vue.js</a></p>
                                <p><a href="http://flask.pocoo.org/">Flask</a></p>
                            </td>
                        </tr>
                        </tbody>
                    </table>
                </div>
            </div>
        </div>
    </div>

    <div id="content2" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">Acknowledgements</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                            <h3 class="text-primary">GISAID data acknowledgement</h3>
                                <span>
                                    We are grateful to the data contributors who shared the data used in this
                                    Web Application via the GISAID Initiative*: the Authors, the Originating
                                    Laboratories responsible for obtaining the specimens, and the
                                    Submitting Laboratories that generated the genetic sequences and metadata.
                                    <br><br>
                                    * Elbe, S., and Buckland-Merrett, G. (2017)
                                    Data, disease and diplomacy: GISAIDâ€™s innovative contribution to global health.
                                    Global Challenges, 1:33-46. DOI: 10.1002/gch2.1018
                                </span>
                            </td>
                        </tr>

                        <tr>
                            <td>
                                <h3 class="text-primary">Other data acknowledgements</h3>
                                EpiSurf includes sequences and metadata from
                                <a href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a>
                                in their Nuccore distribution from the
                                <a href="https://biopython.org/DIST/docs/api/Bio.Entrez-module.html">Entrez APIs.</a>
                                The authors acknowledge the depositions of worldwide laboratories to GenBank and
                                NMDC.
<br><br>
                                EpiSurf also includes sequences and metadata from
                                <a href="https://www.cogconsortium.uk/data/">COG-UK</a>.
                                The authors would like to acknowledge the COG-UK consortium for providing valuable data from across the UK, resulting from the work of hundreds of people. If you wish to publish research using the data, please get in contact with COG-UK first to ensure that those who have generated the data can be involved in its analysis.
                                The COG-UK project is introduced in the
                                <a href="https://www.thelancet.com/action/showPdf?pii=S2666-5247%2820%2930054-9">Consortium paper.</a>

                                <br><br>

                                Epitopes data is imported directly from the Immune Epitope DataBase (<a href="www.iedb.org">IEDB</a>), which should be cited using the <a href="https://doi.org/10.1093/nar/gky1006">most recent publication</a>.
                                We deeply acknowledge its authors for their contribution to the scientific community.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h3 class="text-primary">Personal acknowledgements</h3>
                                The authors would like to thank several researchers
                                for their contribution to requirements elicitation and for inspiring future developments
                                of this research.
                                They are listed in strict alphabetical order by last name.
                                <br>
                                <br>

                                <!--<p style="font-size:20px">Ilaria Capua</p>
                                <p style="font-size:13px">One Health Center of Excellence (University of Florida,
                                    US)</p>
                                <br>-->

                                <p style="font-size:20px">Matteo Chiara</p>
                                <p style="font-size:13px">UniversitÃ  degli Studi di Milano Statale (IT)</p>
                                <br>

                                <!--<p style="font-size:20px">Ana Conesa
                                <p style="font-size:13px">University of Florida (US)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Luca Ferretti
                                <p style="font-size:13px">Oxford Big Data Institute (UK)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Alice Fusaro
                                <p style="font-size:13px">Istituto Zooprofilattico Sperimentale delle Venezie (IT)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Susanna Lamers
                                <p style="font-size:13px">BioInfoExperts (Louisiana, US)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Stefania Leopardi
                                <p style="font-size:13px">Istituto Zooprofilattico Sperimentale delle Venezie (IT)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Francesca Mari
                                <p style="font-size:13px">UniversitÃ  di Siena (IT)</p>
                                <br>-->

                                <p style="font-size:20px">Carla Mavian
                                <p style="font-size:13px">Department of Pathology, College of Medicine (University of
                                    Florida,
                                    US)</p>
                                <br>

                                <!--<p style="font-size:20px">Graziano Pesole
                                <p style="font-size:13px">UniversitÃ  di Bari (IT)</p>
                                <br>-->


                                <!--<p style="font-size:20px">Alessandra Renieri
                                <p style="font-size:13px">UniversitÃ  di Siena (IT)</p>
                                <br>-->

                                <p style="font-size:20px">Brittany Rife Magalis
                                <p style="font-size:13px">Department of Pathology, College of Medicine (University of
                                    Florida,
                                    US)</p>
                                <br>

                                <p style="font-size:20px">Marco Salemi
                                <p style="font-size:13px">Department of Pathology, College of Medicine (University of
                                    Florida,
                                    US)</p>
                                <br>


                                <p style="font-size:20px">Anna Sandionigi
                                <p style="font-size:13px">UniversitÃ  degli Studi di Milano-Bicocca (IT)</p>
                                <br>

                                <!--<p style="font-size:20px">Stephen Tsui
                                <p style="font-size:13px">The Chinese University of Hong Kong (HK)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Limsoon Wong
                                <p style="font-size:13px">National University of Singapore (SGP)</p>
                                <br>-->

                                <!--<p style="font-size:20px">Federico Zambelli
                                <p style="font-size:13px">UniversitÃ  degli Studi di Milano Statale (IT)</p>
                                <br>-->

                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h3 class="text-primary">Resources acknowledgements</h3>
                                We also acknowledge the support from Amazon Machine Learning Research
                                Award "Data-driven Machine and Deep Learning for Genomics".
                            </td>
                        </tr>
                        <!-- <tr>
                            <td>
                                 <h3 class="text-primary">Operational collaborations</h1>
                                The authors would like to thank the following researchers for their technical and knowledge contributions related to the ViruSurf project.
                                <br>
                                <br>

                                <p style="font-size:20px">Tommaso Alfonsi</p>
                                <p style="font-size:13px">Politecnico di Milano (IT)</p>
                                <br>
                                <p style="font-size:20px">Ruba Al Khalaf</p>
                                <p style="font-size:13px">Politecnico di Milano (IT)</p>
                                <br>
                                <p style="font-size:20px">Damianos P. Melidis</p>
                                <p style="font-size:13px">Leibniz University Hannover (DE)</p>
                                <br>
                            </td>
                        </tr>-->
                        </tbody>
                    </table>

                </div>
            </div>
        </div>
    </div>
</div>

<!-- END MAIN PANEL -->
<!-- ==========================CONTENT ENDS HERE ========================== -->

<!-- IMPORTANT: APP CONFIG -->
<script src="js/app.config.js"></script>

<!--[if IE 8]>

<h1>Your browser is out of date, please update your browser by going to www.microsoft.com/download</h1>

<![endif]-->

<!-- Your GOOGLE ANALYTICS CODE Below -->

<!--
<script type="text/javascript">

    var _gaq = _gaq || [];
        _gaq.push(['_setAccount', 'UA-61689851-1']);
        _gaq.push(['_trackPageview']);

    (function() {
        var ga = document.createElement('script');
        ga.type = 'text/javascript';
        ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(ga, s);
    })();

</script>
-->

</body>

</html>

<section id="contact" class="gray-section contact">
    <div class="container">
        <div class="row m-b-lg">
            <div class="col-lg-12 text-center">
                <div class="navy-line"></div>
                <h1>Contact Us</h1>
                <p>For questions, suggestions, bug reports, or any comment</p>
            </div>
        </div>
        <div class="row">
            <div class="col-lg-12 text-center">
                <a href="mailto:arif.canakoglu@polimi.it" class="btn btn-primary">Send us mail</a>
                <br><br>
            </div>
        </div>

    </div>
</section>
