<!DOCTYPE html>
<html lang="en-us">
<head>
    <meta charset="utf-8">
    <!--<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">-->

    <title> About - ViruSurf</title>
    <meta name="description" content="">
    <meta name="author" content="">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <!-- Basic Styles -->
    <link rel="stylesheet" type="text/css" media="screen"
          href="./bootstrap.min.css">
    <link rel="stylesheet" href="inc/font-awesome.min.4.4.0.css">

    <!-- SmartAdmin Styles : Please note (smartadmin-production.css) was created using LESS variables -->
    <link rel="stylesheet" type="text/css" media="screen"
          href="smartadmin-production.min.css">
    <link rel="stylesheet" type="text/css" media="screen" href="contact.css">
    <link rel="icon" href="favicon.ico">

</head>
<body id="extr-page">
<!-- POSSIBLE CLASSES: minified, fixed-ribbon, fixed-header, fixed-width
     You can also add different skin classes such as "smart-skin-1", "smart-skin-2" etc...-->


<div id="main" role="main">
    <!-- MAIN CONTENT -->
    <div id="content0" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">How to cite</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                                <h1 class="text-primary">ViruSurf database and web application</h1>
                                <p >  <font size="3"><b>ViruSurf: an integrated database to investigate viral sequences.</b><br>
                                    A. Canakoglu, P. Pinoli, A. Bernasconi, T. Alfonsi, D.P. Melidis, S. Ceri.
                                    <i>Nucleic Acids Research</i>, gkaa846,
                                    <a href="https://doi.org/10.1093/nar/gkaa846">https://doi.org/10.1093/nar/gkaa846</a></font></p><br>

                                <h1 class="text-primary">Conceptual Model</h1>
                                <p><font size="3"><b>Empowering Virus Sequence Research through Conceptual Modeling.</b>
                                    A. Bernasconi, A. Canakoglu, P. Pinoli, S. Ceri.
                                    In <i>Proceedings of the 39th International Conference on Conceptual Modeling</i>,
                                    November 2020.
                                    <a href="http://www.bioinformatics.deib.polimi.it/geco/publications/Viral_Conceptual_Model___ER2020_GeCoWebsite.pdf">PDF</a></font></p>
                            </td>
                        </tr>
                        </tbody>
                    </table>

                </div>
            </div>
        </div>
    </div>
    <div id="content1" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">Contacts</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>

                                <h1 class="text-primary">Leader</h1>
                                <p class="lead">Prof. Stefano Ceri<br>
                                    <a href="mailto:stefano.ceri@polimi.it">stefano.ceri@polimi.it</a></p>

                                <h1 class="text-primary">Concept and design</h1>

                                <p class="lead">Dr. Arif Canakoglu - Main development<br>
                                    <a href="mailto:arif.canakoglu@polimi.it">arif.canakoglu@polimi.it</a></p>

                                <p class="lead">Dr. Pietro Pinoli - Variant algorithms<br>
                                    <a href="mailto:pietro.pinoli@polimi.it">pietro.pinoli@polimi.it</a></p>

                                <p class="lead">Anna Bernasconi - Configuration, testing, documentation<br>
                                    <a href="mailto:anna.bernasconi@polimi.it">anna.bernasconi@polimi.it</a>

                                <address>
                                    <strong>Dipartimento di Elettronica, Informazione e Bioingegneria</strong><br>
                                    <strong>Politecnico di Milano</strong><br>
                                    Via Ponzio 34/5 Milano<br>
                                    20133 Milano<br>
                                    Italy<br>
                                    <br>
                                    Phone: +39 02 2399 3655
                                </address>
                            </td>
                        </tr>


                        <tr>
                            <td>
                                <h1 class="text-primary">Third-party software</h1>
                                <p><a href="https://www.postgresql.org/">PostgreSQL</a></p>
                                <p><a href="https://vuejs.org/">Vue.js</a></p>
                                <p><a href="http://flask.pocoo.org/">Flask</a></p>
                            </td>
                        </tr>
                        </tbody>
                    </table>

                </div>
            </div>
        </div>
    </div>
    <div id="content2" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">Acknowledgements</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                                <h1 class="text-primary">Available data</h1>
                                ViruSurf includes sequences and metadata from
                                <a href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a>
                                in their Nuccore distribution from the
                                <a href="https://biopython.org/DIST/docs/api/Bio.Entrez-module.html">Entrez APIs.</a>
                                <br>
                                It also includes sequences and metadata from
                                <a href="https://www.cogconsortium.uk/data/">COG-UK</a>
                                in the last publicly available distribution of 2020-05-08.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Data acknowledgements</h1>
                                The authors would like to acknowledge the COG-UK consortium for providing valuable data
                                from across the UK,
                                resulting from the work of hundreds of people. If you wish to publish research using the
                                data,
                                please get in contact with COG-UK first to ensure that those who have generated the data
                                can be involved in its analysis.
                                <br>
                                The COG-UK project is introduced in the
                                <a href="https://www.thelancet.com/action/showPdf?pii=S2666-5247%2820%2930054-9">Consortium
                                    paper.</a>
                                <br><br>
                                The authors also acknowledge the depositions of worldwide laboratories to GenBank and
                                NMDC.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Personal acknowledgements</h1>
                                The authors would like to thank several researchers
                                for their contribution to requirements elicitation and for inspiring future developments
                                of this research.
                                They are listed in strict alphabetical order by last name.
                                <br>
                                <br>

                                <p style="font-size:20px">Ilaria Capua</p>
                                <p style="font-size:13px">One Health Center of Excellence (University of Florida,
                                    US)</p>
                                <br>

                                <p style="font-size:20px">Matteo Chiara</p>
                                <p style="font-size:13px">Universit√† degli Studi di Milano Statale (IT)</p>
                                <br>

                                <p style="font-size:20px">Ana Conesa
                                <p style="font-size:13px">University of Florida (US)</p>
                                <br>

                                <p style="font-size:20px">Luca Ferretti
                                <p style="font-size:13px">Oxford Big Data Institute (UK)</p>
                                <br>

                                <p style="font-size:20px">Alice Fusaro
                                <p style="font-size:13px">Istituto Zooprofilattico Sperimentale delle Venezie (IT)</p>
                                <br>

                                <p style="font-size:20px">Ruba Al Khalaf
                                <p style="font-size:13px">Politecnico di Milano (IT)</p>
                                <br>

                                <p style="font-size:20px">Susanna Lamers
                                <p style="font-size:13px">BioInfoExperts (Louisiana, US)</p>
                                <br>

                                <p style="font-size:20px">Stefania Leopardi
                                <p style="font-size:13px">Istituto Zooprofilattico Sperimentale delle Venezie (IT)</p>
                                <br>

                                <p style="font-size:20px">Francesca Mari
                                <p style="font-size:13px">Universit√† di Siena (IT)</p>
                                <br>

                                <p style="font-size:20px">Carla Mavian
                                <p style="font-size:13px">Department of Pathology, College of Medicine (University of
                                    Florida,
                                    US)</p>
                                <br>

                                <p style="font-size:20px">Graziano Pesole
                                <p style="font-size:13px">Universit√† di Bari (IT)</p>
                                <br>


                                <p style="font-size:20px">Alessandra Renieri
                                <p style="font-size:13px">Universit√† di Siena (IT)</p>
                                <br>

                                <p style="font-size:20px">Anna Sandionigi
                                <p style="font-size:13px">Universit√† degli Studi di Milano-Bicocca (IT)</p>
                                <br>

                                <p style="font-size:20px">Stephen Tsui
                                <p style="font-size:13px">The Chinese University of Hong Kong (HK)</p>
                                <br>

                                <p style="font-size:20px">Limsoon Wong
                                <p style="font-size:13px">National University of Singapore (SGP)</p>
                                <br>

                                <p style="font-size:20px">Federico Zambelli
                                <p style="font-size:13px">Universit√† degli Studi di Milano Statale (IT)</p>
                                <br>

                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Resources acknowledgements</h1>
                                We also acknowledge the support from Amazon Machine Learning Research
                                Award "Data-driven Machine and Deep Learning for Genomics".
                            </td>
                        </tr>
                        <!-- <tr>
                            <td>
                                 <h1 class="text-primary">Operational collaborations</h1>
                                The authors would like to thank the following researchers for their technical and knowledge contributions related to the ViruSurf project.
                                <br>
                                <br>

                                <p style="font-size:20px">Tommaso Alfonsi</p>
                                <p style="font-size:13px">Politecnico di Milano (IT)</p>
                                <br>
                                <p style="font-size:20px">Ruba Al Khalaf</p>
                                <p style="font-size:13px">Politecnico di Milano (IT)</p>
                                <br>
                                <p style="font-size:20px">Damianos P. Melidis</p>
                                <p style="font-size:13px">Leibniz University Hannover (DE)</p>
                                <br>
                            </td>
                        </tr>-->
                        </tbody>
                    </table>

                </div>
            </div>
        </div>
    </div>
</div>

<!-- END MAIN PANEL -->
<!-- ==========================CONTENT ENDS HERE ========================== -->

<!-- IMPORTANT: APP CONFIG -->
<script src="js/app.config.js"></script>

<!--[if IE 8]>

<h1>Your browser is out of date, please update your browser by going to www.microsoft.com/download</h1>

<![endif]-->

<!-- Your GOOGLE ANALYTICS CODE Below -->

<!--
<script type="text/javascript">

    var _gaq = _gaq || [];
        _gaq.push(['_setAccount', 'UA-61689851-1']);
        _gaq.push(['_trackPageview']);

    (function() {
        var ga = document.createElement('script');
        ga.type = 'text/javascript';
        ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(ga, s);
    })();

</script>
-->

</body>

</html>

<section id="contact" class="gray-section contact">
    <div class="container">
        <div class="row m-b-lg">
            <div class="col-lg-12 text-center">
                <div class="navy-line"></div>
                <h1>Contact Us</h1>
                <p>For questions, suggestions, bug reports, or any comment</p>
            </div>
        </div>
        <div class="row">
            <div class="col-lg-12 text-center">
                <a href="mailto:arif.canakoglu@polimi.it" class="btn btn-primary">Send us mail</a>
                <br><br>
            </div>
        </div>

    </div>
</section>
