<!DOCTYPE html>
<html lang="en-us">
<head>
    <meta charset="utf-8">
    <!--<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">-->

    <title> Data curation - ViruSurf</title>
    <meta name="description" content="">
    <meta name="author" content="">

    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <!-- Basic Styles -->
    <link rel="stylesheet" type="text/css" media="screen"
          href="./bootstrap.min.css">
    <link rel="stylesheet" href="inc/font-awesome.min.4.4.0.css">

    <!-- SmartAdmin Styles : Please note (smartadmin-production.css) was created using LESS variables -->
    <link rel="stylesheet" type="text/css" media="screen"
          href="smartadmin-production.min.css">
    <link rel="stylesheet" type="text/css" media="screen" href="contact.css">
    <link rel="icon" href="favicon.ico">

</head>
<body id="extr-page">
<!-- POSSIBLE CLASSES: minified, fixed-ribbon, fixed-header, fixed-width
     You can also add different skin classes such as "smart-skin-1", "smart-skin-2" etc...-->


<div id="main" role="main">
    <!-- MAIN CONTENT -->
    <div id="content" class="container">
        <div class="row">
            <div class="col-sm-6">
                <div class="bs-example bs-example-type">
                    <h2 class="text-primary" style="margin: 20px 0; font-size:38px">Data curation</h2>
                    <table class="table" style="background-color: #fafafa;">
                        <tbody>
                        <tr>
                            <td>
                                <h1 class="text-primary">Curation principles</h1>
                                Data must be captured, standardized, organized, and made accessible to the scientific
                                community in order to have a significant and lasting impact.
                                The role of ViruSurf is to provide up-to-date, accurate, and accessible information,
                                and, through this critical activity, facilitate virology scientific discovery.
                            </td>
                        </tr>

                        <tr>
                            <td>
                                <h1 class="text-primary">Metadata curation</h1>
                                Our first curation contribution is to provide a unique schema
                                (see <a href="https://www.biorxiv.org/content/10.1101/2020.04.29.067637v3">Viral
                                Conceptual Model - VCM</a>) for different data sources.
                                Different ad-hoc importer modules have been built to load GenBank, RefSeq, COG-UK,
                                and GISAID data; the code is available <a
                                    href="https://github.com/DEIB-GECO/virusurf_downloader">here</a>.
                                Each source comes with different terminological choices when describing the different
                                metadata.
                                We have surveyed the the terms used in many sources to come up with appropriate
                                reconciling solutions.
                                <!--<br><br>
                                <img src="./table.jpg" alt="table to compare terminology">
                                <br><br>
                                A complete mapping of our terms with the integrated sources can be found at...-->
                                All import results have been browsed and manually checked by the development team.
                                <br><br>
                                Note that for the <a href="http://gmql.eu/virusurf_gisaid">GISAID-specific interface of
                                ViruSurf</a>
                                we use attribute names as specifie din GISAID EpiCoV database. In case the reconciling
                                terminology from VCM is different, it is reported within parentheses in second position.
                                <br><br>
                                Specific curation efforts have been dedicated to:
                                <br>
                                <ul>
                                    <li>
                                        <b>Location</b> information.
                                        GISAID provides the Location; we split the field according to the separator "/"
                                        and save the first three levels
                                        into the GeoGroup, the Country, and the Region.
                                        GenBank provides always the Country, often the Region almost always, while
                                        the GeoGroup is added from a maintained
                                        <a href="https://github.com/DEIB-GECO/virusurf_downloader/blob/master/geo_groups.py">
                                            map of continents-country</a>.
                                        In COG-UK we retrieve relevant location information and complement Region and
                                        GeoGroup.
                                    </li>
                                    <li>
                                        <b>Collection date</b>.
                                        GISAID is reported as communicated by GISAID team.
                                        For other sources, to make sequences more easily searchable through date
                                        filters,
                                        when the specific day is not available, we typically add 01 (the first day of
                                        the month)
                                        when the specific month is not available, we use default date (01 day, 01
                                        month).
                                        If the date is missing in general, we check the referenced publication and use
                                        the provided information.
                                    </li>
                                    <li>
                                        <b>Submission date</b>.
                                        GISAID provides the field.
                                        In GenBank it is retrieved from the JOURNAL field for SARS-CoV2, while at times
                                        from the PATENT field for SARS-CoV entries.
                                        In COG-UK we set it to null.
                                        Curation rules hold in the same way as for collection date.
                                    </li>
                                    <li>
                                        <b>Virus taxon name</b>, <b>taxon ID</b>, and <b>species</b> are critical fields
                                        as they depend on the
                                        information retrieved through the authoritative NCBI Taxonomy.
                                        We make a direct call to Entrez APIs:
                                        <p style="font-family:courier;">
                                            Entrez.fetch(db="taxonomy",id=taxon_id,rettype=None,retmode="xml")</p>
                                        However, some viruses are described as "species"
                                        (such as SARS-CoV), some others as "sub-species" (such as SARS-CoV 2).
                                        This may result in confusion, e.g., when comparing SARS-CoV vs. SARS-CoV 2.
                                        <br>
                                        We have thus chosen a comprehensive approach:
                                        the field taxon name refers to the most specific term used to define the type of
                                        virus under examination, while the field taxon ID is the related numerical
                                        identifier withing NCBI Taxon.
                                        Instead, Species indicates the direct parent (in the taxonomy graph) of the
                                        taxon name (when this is not a "species" according to the taxonomy)
                                        or the taxon name itself (when it is a "species").
                                    </li>
                                    <li>
                                        As to the host organism information, we also add the <b>Host taxon ID</b> (which
                                        complements the <b>Host taxon name</b>)
                                        manually; in some cases we do not find exact match with NCBI Taxonomy entries so
                                        we make our translation,
                                        e.g., "canine" becomes "Canis lupus familiaris".
                                        <br>
                                        In some rare cases, sequences come with gender and age information, which is
                                        parsed directly from the "host"-related string
                                        (e.g., host="Homo sapiens; male; age 65").
                                    </li>
                                    <li>
                                        From COG-UK and GISAID we only import SARS-CoV-2 data. For this kind of virus
                                        the <b>reference sequence</b>
                                        is clear.
                                        From GenBank, not all the species of virus have a clearly specified reference
                                        sequence. We thus define the reference sequence
                                        by previously cross-checking with several research papers to ascertain that the
                                        typical reference sequence used for variant calling is defined.
                                    </li>
                                    <li>
                                        It is not always clear how to define a sequence <b>complete</b> or
                                        <b>partial</b>.
                                        From GISAID this information is imported.
                                        For others, to set the "complete" value, in GenBank we look for the keyword
                                        "complete
                                        genome" in the DEFINITION field.
                                        For "partial" value, instead, we look for "partial CDS", "complete cds" or
                                        "partial genome".
                                        Otherwise (and alwaysin COG-UK), we compute this information according to
                                        the following rule: if the
                                        sequence length is >= 95% of the reference length then
                                        Is Complete = True, else Is Complete = false.
                                    </li>
                                    <li>
                                        <b>Strand</b>, as well as <b>Molecule Type</b>, <b>Is Single Stranded</b> and
                                        <b>Is Positive Stranded</b>
                                        are curated manually.
                                    </li>
                                    <li>
                                        <b>GC%</b> and <b>N%</b> are provided by GISAID for their data, while
                                        have been computed to provide users with an additional
                                        information on
                                        the quality of the sequence.
                                    </li>
                                    <li>
                                        The information on <b>sequencing technology</b>, <b>assembly method</b> and <b>coverage</b>
                                        is only available from GenBank.
                                        We look for a specific pattern in the COMMENT field (i.e.,
                                        ##Assembly-Data-START## < content > ##Assembly-Data-END##
                                        <br>
                                        For the Coverage field, a series of curation steps were necessary,
                                        as the values gathered from GenBank were very heterogeneous in the notation.
                                        <ul>
                                            <li>remove strings (possibly within parenthesis) that on the left/right of
                                                the numbers (including x, X, reads/nt, >, <.....)
                                            </li>
                                            <li>commas or dots in between numbers should be treated in two different
                                                ways:
                                                <ul>
                                                    <li>if they have three numbers on their right: remove</li>
                                                    <li>if they have less than three numbers on their right (so 1 or 2),
                                                        treat those numbers as decimals and round the unit to the
                                                        closest
                                                        one (e.g. 112593.3 becomes 112593, whilst 70.8 becomes 71)
                                                    </li>
                                                </ul>
                                            </li>
                                        </ul>
                                        These rules were applied after carefully checking many examples in the originary
                                        research papers to
                                        ascertain that our homogenization functions were consistent with the original
                                        semantics.
                                    </li>
                                </ul>
                                <br><br>
                                As an additional curation, the records of COG-UK are integrated with 5 additional metadata
                                (Originating lab, Submission date, Submission Lab, Isolation source, IsComplete) when they correspond
                                to GISAID records that hold such information.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Data curation</h1>
                                Data curation is not performed for instances of GISAID, as we do not import sequences
                                from this source.
                                Amino acid level changes are imported from periodical exports from GISAID and simply
                                parsed into
                                our database.

                                For other sources, we engineered our pipelines to provide a unique annotation procedure
                                that will seamlessly process sequences that come with existing annotations as well as
                                sequences that do not.
                                <br>
                                This process is highly critical for being able to supply users with comparable
                                information.
                                <br>
                                <br>
                                For the portion of already annotated sequences (i.e., from GenBank),
                                our in-house algorithm compares its results with the existing annotations and, when
                                different these spot cases are manually re-checked to ensure correctness.

                                We extract:
                                structural annotations (filling the \textsc{Annotation} table),
                                nucleotide and amino acid sequences for each annotated segment,
                                nucleotide variants and their impact,
                                amino acid variants for the proteins,
                                other information such as percentage of specific nucleotide bases.
                                <br><br>
                                In order to compute such pieces of information, for each virus we manually select a
                                reference sequence and a set of annotations, comprising coordinates for codifying and
                                structural regions, as well as the amino acid sequences of each protein. Usually, such
                                data are taken from the RefSeq entry for the given virus (e.g., NC_045512 for
                                SARS-CoV2), yet our import process is not bound to RefSeq and any source that can
                                provide such information may be used.
                                <br><br>
                                For each imported sequence, the pipeline starts by computing the optimal global
                                alignment to the reference by means of the <a
                                    href="https://www.sciencedirect.com/science/article/abs/pii/0022283670900574">
                                dynamic programming Needleman-Wunsch (NW)
                                algorithm</a>.
                                The time and space complexity of NW is quadratic
                                in the length of the aligned sequences, which often hinders its adoption in real
                                scenarios; however, in the virus case the sequences are relative small and we preferred
                                to use NW rather than faster heuristic techniques to ensure that the optimal alignment
                                is found for every sequence. Moreover, we configure the algorithm to use an affine gap
                                penalty, so as to favor longer gaps which are very frequent at the ends of sequences.
                                <br><br>
                                Once the alignment is computed, all the differences (variants) with the reference
                                sequence are collected in the form of nucleotide substitutions, insertions or deletions.
                                Using the <a href="http://snpeff.sourceforge.net/">SnpEff tool</a> (see
                                <a href="http://snpeff.sourceforge.net/SnpEff_manual.html">SnpEff documentation</a>)
                                we annotate each variant and predict its impact
                                on the codifying regions; indeed, a variant may, for example, be irrelevant (e.g, when
                                the mutated codon codifies for the same amino acid of the original codon), produce small
                                changes, or be deleterious.
                                Based on the alignment result, the sub-sequences corresponding to the reference
                                annotations are identified within the input sequence.
                                <br>
                                <br>
                                Coding regions are translated into their equivalent amino acid sequences. The
                                translation mechanism takes into consideration annotated ribosomial frameshifts events
                                (e.g., within the ORF1ab gene of SARS-CoV2) to compute the correct amino acid sequence.
                                When the translation fails (e.g., because the nucleotide sequence retrieved from the
                                alignment is empty or its length is not a multiple of 3), we ignore such amino acid
                                product; indeed, such events are mainly due to incompleteness and poor quality of the
                                input sequence and further investigating these sequences for amino acid variants would
                                lead to erroneous information.
                                Instead, when an aligned codon contains any IUPAC character ambiguously representing a
                                set of bases <a href="https://genome.ucsc.edu/goldenPath/help/iupac.html">(https://genome.ucsc.edu/goldenPath/help/iupac.html)</a>, it is
                                translated into the X (unknown) amino acid, which automatically becomes a
                                variant.
                                Notice that queries selecting known amino acids work correctly independently from
                                unknown ones (that in most cases will not be of interest).
                                Otherwise, the translated amino acid sequence is further
                                aligned -- again using NW -- with the corresponding amino acid sequence, annotated with
                                the reference, and its eventual amino acid variants are inferred.
                                <br>
                                The sequences provided by COG-UK are already aligned to the reference genome. However,
                                our pipeline firstly reverts the alignment (by simply removing the inserted gaps) and
                                then recomputes it using the same reference sequence and alignment algorithm used for
                                other sources. This choice has been done to ensure uniformity of the alignment among
                                different data sources.
                                <br>
                                <br>
                                Note that for protein names we use the typical terminology of GISAID, while we put
                                within parentheses
                                in second position the (possibly) alternative terminology adopted by NBI GenBank-related
                                resources.
                            </a>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Overlap detection and management</h1>
                                ViruSurf solves the problem of record redundancy among different databases by using
                                different external references IDs that are either available from the origin sources or
                                by computing in-house the matches between strain names and sequence length.
                                We store this information and allow the possibility of performing queries on just
                                ``GISAID only'' sequences, i.e., that are not present in GenBank, COG-UK or NMDC.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Case/control computation</h1>
                                The `Show control' switch allows to visualize the sequences of the control group,
                                defined by those sequences selected by the Metadata search filters for which the Variant
                                filters are not satisfied. Note that, for some sequences amino acid sequences could not
                                be
                                computed (e.g., because the nucleotide sequence retrieved from the alignment is empty or
                                its length is not a multiple of 3).
                                The "controls" exclude such sequences.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <h1 class="text-primary">Reference sequences</h1>
                                <ul>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_045512.2">NC_045512.2 </a> for
                                        SARS-CoV-2
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_004718.3">NC_004718.3 </a> for
                                        SARS-CoV
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_019843.3">NC_019843.3 </a> for
                                        MERS-CoV
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_000000.0">NC_000000.0 </a> for
                                        Bat SARS-like coronavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_038294.1">NC_038294.1 </a> for
                                        Betacoronavirus England 1
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_039345.1">NC_039345.1 </a> for
                                        Bombali ebolavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_014373.1">NC_014373.1 </a> for
                                        Bundibugyo ebolavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_004161.1">NC_004161.1 </a> for
                                        Reston ebolavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_006432.1">NC_006432.1 </a> for
                                        Sudan ebolavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_014372.1">NC_014372.1 </a> for
                                        Tai Forest ebolavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_002549.1">NC_002549.1 </a> for
                                        Zaire ebolavirus
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_001477.1">NC_001477.1 </a> for
                                        Dengue virus 1
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_001474.2">NC_001474.2 </a> for
                                        Dengue virus 2
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_001475.2">NC_001475.2 </a> for
                                        Dengue virus 3
                                    </li>
                                    <li><a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_002640.1">NC_002640.1 </a> for
                                        Dengue virus 4
                                    </li>
                                </ul>
                            </td>
                        </tr>
                        </tbody>
                    </table>

                </div>
            </div>
        </div>
    </div>
</div>

<!-- END MAIN PANEL -->
<!-- ==========================CONTENT ENDS HERE ========================== -->

<!-- IMPORTANT: APP CONFIG -->
<script src="js/app.config.js"></script>

<!--[if IE 8]>

<h1>Your browser is out of date, please update your browser by going to www.microsoft.com/download</h1>

<![endif]-->

<!-- Your GOOGLE ANALYTICS CODE Below -->

<!--
<script type="text/javascript">

    var _gaq = _gaq || [];
        _gaq.push(['_setAccount', 'UA-61689851-1']);
        _gaq.push(['_trackPageview']);

    (function() {
        var ga = document.createElement('script');
        ga.type = 'text/javascript';
        ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0];
        s.parentNode.insertBefore(ga, s);
    })();

</script>
-->

</body>

</html>

<section id="contact" class="gray-section contact">
    <div class="container">
        <div class="row m-b-lg">
            <div class="col-lg-12 text-center">
                <div class="navy-line"></div>
                <h1>Contact Us</h1>
                <p>For questions, suggestions, bug reports, or any comment</p>
            </div>
        </div>
        <div class="row">
            <div class="col-lg-12 text-center">
                <a href="mailto:arif.canakoglu@polimi.it" class="btn btn-primary">Send us mail</a>
                <br><br>
            </div>
        </div>

    </div>
</section>
